osgp/platform/osgp-adapter-kafka-distributionautomation/src/main/java/org/opensmartgridplatform/adapter/kafka/da/application/mapping/MeterReadingMapping.java

```diff
@@ -0,0 +1,115 @@
+import org.opensmartgridplatform.adapter.kafka.ReadingType;
+import org.opensmartgridplatform.adapter.kafka.UsagePoint;
+import org.opensmartgridplatform.adapter.kafka.ValuesInterval;
+import org.opensmartgridplatform.domain.da.measurements.Measurement;
+import org.opensmartgridplatform.domain.da.measurements.MeasurementElement;
+import org.opensmartgridplatform.domain.da.measurements.MeasurementGroup;
+import org.opensmartgridplatform.domain.da.measurements.MeasurementReport;
+import org.opensmartgridplatform.domain.da.measurements.elements.FloatMeasurementElement;
+import org.opensmartgridplatform.domain.da.measurements.elements.TimestampMeasurementElement;
+
+import ma.glasnost.orika.CustomConverter;
+import ma.glasnost.orika.MappingContext;
+import ma.glasnost.orika.metadata.Type;
+
+public class MeterReadingMapping extends CustomConverter<MeasurementReport, MeterReading> {
+
+    @Override
+    public MeterReading convert(final MeasurementReport source, final Type<? extends MeterReading> destinationType,
+            final MappingContext mappingContext) {
+
+        final List<MeasurementElement> measurementElements = source.getMeasurementGroups()
+                .stream()
+                .map(MeasurementGroup::getMeasurements)
+                .flatMap(List::stream)
+                .map(Measurement::getMeasurementElements)
+                .flatMap(List::stream)
+                .collect(Collectors.toList());
+
+        final List<IntervalReading> intervalReadings = Arrays.asList(this.getIntervalReading(measurementElements));
+
+        final String identification = source.getMeasurementGroups()
+                .get(0)
+                .getIdentification();
+        final ReadingType readingType = this.getReadingType(identification);
+        final IntervalBlock intervalBlock = new IntervalBlock(readingType, intervalReadings);
+
+        final Long start = intervalReadings.stream()
+                .map(IntervalReading::getTimeStamp)
+                .min(Long::compareTo)
+                .orElse(dateStringToEpoch("2020-01-13 00:00:00"));
+        final Long end = intervalReadings.stream()
+                .map(IntervalReading::getTimeStamp)
+                .max(Long::compareTo)
+                .orElse(dateStringToEpoch("2020-01-16 00:00:00"));
+        final ValuesInterval valuesInterval = new ValuesInterval(start, end);
+
+        final UsagePoint usagePoint = new UsagePoint(this.getUsagePoint(identification));
+        return new MeterReading(valuesInterval, null, identification, usagePoint, Arrays.asList(intervalBlock));
+
+    }
+
+    private ReadingType getReadingType(final String identification) {
+        if (identification.contains("Power")) {
+            return new ReadingType("10s", "energy", "M", "W", "Energy MW en 10s", null);
+        } else {
+            return new ReadingType("60s", "temperature", "none", "degC", "Temperature transformateur °C en 60s", null);
+        }
```

Comment:
This code was good enough for the POC. I doubt if it is good enough now.

Reply:
Leaving for now, will become obsolete when new AVRO message format is defined